db/db_compaction_test.cc

```diff
@@ -6278,327 +6273,709 @@ void IngestOneKeyValue(DBImpl* db, const std::string& key,
+               compaction_path_to_test == "CompactRange") {
+      assert(options_.compaction_style ==
+             CompactionStyle::kCompactionStyleFIFO);
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+          "PickCompactionToWarm::BeforeGetCurrentTime",
+          [this](void* current_time_arg) -> void {
+            // The unit test goes so quickly that there is almost no time
+            // elapsed after we ingest a file and before we check whether
+            // ingested files can compact to warm. Therefore we need this trick
+            // to simulate elapsed time in reality.
+            int64_t* current_time = (int64_t*)current_time_arg;
+            *current_time = *current_time +
+                            options_.compaction_options_fifo.age_for_warm + 1;
+          });
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+          "PostPickCompactionToWarm", [&](void* compaction_arg) {
+            Compaction* compaction = (Compaction*)compaction_arg;
+            if (compaction != nullptr) {
+              compaction_path_sync_point_called_.store(true);
+            }
+          });
+    }
+
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+  }
+
+  bool SyncPointsCalled() { return compaction_path_sync_point_called_.load(); }
+
+  void DisableSyncPoints() {
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->ClearAllCallBacks();
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
+  }
+
+  // For files of largest seqno less than `strict_largest_seqno_uppper_bound`,
+  // return the largest seqno of the latest L0 file (based on file number)
+  SequenceNumber GetLatestL0FileLargestSeqnoHelper(
+      SequenceNumber strict_largest_seqno_uppper_bound = kMaxSequenceNumber) {
+    VersionSet* const versions = dbfull()->GetVersionSet();
+    assert(versions);
+    ColumnFamilyData* const cfd = versions->GetColumnFamilySet()->GetDefault();
+    assert(cfd);
+    Version* const current = cfd->current();
+    assert(current);
+    VersionStorageInfo* const storage_info = current->storage_info();
+    assert(storage_info);
+    const std::vector<FileMetaData*> level0_files = storage_info->LevelFiles(0);
+    assert(level0_files.size() >= 1);
+
+    uint64_t latest_file_num = 0;
+    uint64_t latest_file_largest_seqno = 0;
+    for (FileMetaData* f : level0_files) {
+      if (f->fd.largest_seqno < strict_largest_seqno_uppper_bound &&
+          f->fd.GetNumber() > latest_file_num) {
+        latest_file_num = f->fd.GetNumber();
+        latest_file_largest_seqno = f->fd.largest_seqno;
+      }
+    }
+
+    return latest_file_largest_seqno;
+  }
+
+ protected:
+  Options options_;
+
+ private:
+  const Snapshot* snapshot_ = nullptr;
+  std::atomic<bool> compaction_path_sync_point_called_;
+  std::shared_ptr<test::SleepingBackgroundTask> sleeping_task_;
+};
+
+TEST_F(DBCompactionTestL0FilesMisorderCorruption,
```

Comment:
[Note] I combine the previous level compaction's test FlushAfterIntraL0CompactionCheckConsistencyFail and IntraL0CompactionAfterFlushCheckConsistencyFail cuz one test is enough to make the point now.